### DISCUSSION TITLE
When are API references upgraded?

### DISCUSSION THREAD

**Discussion by yildizbilal** [fluxcd/flux2#2972] [General] [answered]
I just want to learn how flux works. And i am a bit confused.

I wanted to upgrade the api references of the flux components - never did that before. So i changed the apiVersion of a source from `source.toolkit.fluxcd.io/v1beta1` to `source.toolkit.fluxcd.io/v1beta2` and pushed it to git. I watched the cluster with k9s because i didn't know what will happen. Furthermore i saw that everything reconciled properly, so i checked the apiVersion of the `HelmRepository` object whose version i changed and the apiVersion was set correctly to the new version.

Then i recognised that all `HelmRepository` objects has the current apiVersion already. And other used CRDs by `toolkit.fluxcd` also. But the change doesn't come from the manifest files, so our git repository manifest files doesn't map the cluster correctly. 

### **When and what updated the apiVersions of all `toolkit.fluxcd` CRDs automatically?**

Recently i upgraded the flux components by overwriting the `gotk-components.yaml` file with the `flux install ..` command. Could that cause the apiVersion changes automatically? Where can i find an explanation about this? 

I found in the discussion feeds some entries about how to upgrade from one specific api version to another. Have somebody a link reference for something about `How to upgrade to a new api reference in general?`. I don't understood that yet und couldn't find good articles about this topic.

Thanks.

**Comment by stefanprodan** [comment, marked as answer]
Kubernetes upgrades the custom resources automatically when the custom resource definitions gets updated. The CRDs are included in `gotk-components.yaml`.  After the CRDs upgrade, it is recommended to bump the API version in Git, failing to do so will result in various bugs depending on how old your Kubernetes version is.

You can read more about CR and CRDs here https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

**Reply by yildizbilal** [reply]
I don't understood why they upgraded automatically - could that not raise other problems? But thanks for you reply 🙏. I'll bump the current in the cluster running API versions of the flux CRDs to our git, so that our manifest files maps the cluster correctly.

**Reply by stefanprodan** [reply]
> I don't understood why they upgraded automatically - could that not raise other problems?

This is how Kubernetes works, once you update the CRDs, all in-cluster CRs are automatically upgraded to the latest versions of the CRDs. We have extensive e2e tests in place to make sure the upgrade doesn't break anything. If manual changes are required to CRs, we publish the upgrade procedure in the release notes.
